File: app/src/main/java/com/example/usermanagmentapp/data/repository/UserRepository.java
package com.example.usermanagmentapp.data.repository;

import com.example.usermanagmentapp.data.model.User;
import com.example.usermanagmentapp.data.model.UserStatus;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;

public interface UserRepository {
    /**
     * Get a list of users from the a source.
     *
     * @param page    Page number
     * @param perPage Number of users per page
     * @return [List] of [User] or an error
     */
    Single<List<User>> users(int page, int perPage);

    /**
     * Get a user from the a source.
     *
     * @param id User id
     * @return [User]
     */
    Single<User> user(int id);

    /**
     * Add a new user to the a source.
     *
     * @param name   name
     * @param gender gender
     * @param email  email
     * @param status status
     * @return [Completable] that return no value
     */
    Completable addUser(String name, String gender, String email, UserStatus status);

}

File: app/src/main/java/com/example/usermanagmentapp/data/repository/UserRepository.kt
package com.example.usermanagmentapp.data.repository

import com.example.usermanagmentapp.data.UserNotFoundError

import com.example.usermanagmentapp.data.UsersFetchError
import com.example.usermanagmentapp.data.model.User
import com.example.usermanagmentapp.data.model.UserStatus
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single
import kotlin.jvm.Throws

/**
 * Entry point for accessing users data.
 */
interface UserRepository {
    /**
     * Get a list of users from the a source.
     * @param page Page number
     * @param perPage Number of users per page
     * @return [List] of [User] or an error
     */
    fun users(page: Int, perPage: Int): Single<List<User>>

    /**
     * Get a user from the a source.
     * @param id User id
     * @return [User]
     */
    fun user(id: Int): Single<User>

    /**
     * Add a new user to the a source.
     * @param name
     * @param gender
     * @param email
     * @param status
     * @return [Completable] that return no value
     */
    fun addUser(name: String, gender: String, email: String, status: UserStatus): Completable

}